### ISSUE TITLE
this.elemParams на ноде с двумя элементами

### ISSUE THREAD

**Issue by h4** [bem/bem-forum-content-ru#579]
Есть блок `post`, у него есть элемент `<span class="post__action post__create" data-bem='{"post__create": {"url": "/post/create"} }'>Создать</span>`.

На `post__create` навешен обработчик клика. Примерно такой:
```js
_sendCreateRequest: function(e) {
    var $target = $(e.target),
        params = this.elemParams($target);

   $.get(params.url)
        .done(function() {});
}
```

И вот будет этот  обработчик работать правильно или нет — зависит от порядка, в котором обявлены элементы на дом-ноде. `<span class="post__action post__create">` — не работает, `<span class="post__create post__action">` — работает.

И вот вопрос — это косяк в логике блока и так делать ообще не нужно никогда, или всё-таки баг в `this.elemParams()` и его стоит зарепортить?

**Comment by sipayRT** [comment]
Стоит исправить логику. На одной дом-ноде не должно быть двух элементов блока. Представь что будет, когда ты для обоих элементов напишешь шаблоны - что ты будешь ожидать в результате?

**Comment by apsavin** [comment]
На одной дом-ноде вполне может быть два элемента блока, если второй миксуется к первому.

**Comment by sipayRT** [comment]
да, при миксе ты можешь добавить элемент другого блока. Но какой смысл в миксе в пределах одного блока?

**Comment by zxqfox** [comment]
Точно такой же, как и при миксе 2х блоков на одной ноде.

Банально, если в одном случае кнопку с действием и какой-то индикатор нам надо разделять, а в другом — в одном месте рисовать. В одном случае это две разные дом-ноды, в другом — одна.

**Comment by h4** [comment]
Ну и да, добавлю — у нас half-стек, шаблоны на twig/handlebars.

**Comment by zxqfox** [comment]
Я, конечно, должен предложить конвертировать twig/handlebars в bh/bh-php ;-)

**Comment by sipayRT** [comment]
> Банально, если в одном случае кнопку с действием и какой-то индикатор нам надо разделять, а в другом — в одном месте рисовать. В одном случае это две разные дом-ноды, в другом — одна.

Давай по-пацански:

У меня в голове сейчас именно такая картинка. Есть блок `a`, который у дочернего элемента спрашивает "ты кто?!", а ему отвечают я элемент `b`, но со мной тут еще ребята намиксованы. Т.е. сущность все равно одна, пофиг сколько намиксовано всего. Но должен оставаться кто-то, кто будет "за главного" :)

**Comment by sipayRT** [comment]
> Я, конечно, должен предложить конвертировать twig/handlebars в bh/bh-php ;-)

обязан ;)

**Comment by zxqfox** [comment]
@sipayRT это нужно для шаблонов, потому что там конфликт интересов, который не придумали как разрешать. Для JS мы легко можем создать столько инстансов, сколько надо, и когда надо, я не вижу проблемы иметь несколько инстансов, привязанных к одной дом-ноде. Равно как и в CSS. И тут, и там главный уже не нужен, главный нужен только для шаблонов.

**Comment by h4** [comment]
@sipayRT только вот почему-то ответ на вопрос «кто главный» сделан по принципу — кто «последний, тот и папа» (ну, то есть наоборот, кто первый).

**Comment by sipayRT** [comment]
> только вот почему-то ответ на вопрос «кто главный» сделан по принципу — кто «последний, тот и папа» (ну, то есть наоборот, кто первый).

вот у меня в голове именно такой принцип.

> я не вижу проблемы иметь несколько инстансов, привязанных к одной дом-ноде

я не могу прям сейчас придумать такую ситуацию, которая убедила бы тебя в моей правоте :(

**Comment by apsavin** [comment]
Мне кажется, не важно, стоит смешивать элементы блока на ноде или нет. Важно, что это возможно и что в такой ситуации поведение elemParams как минимум не очевидно/не документировано, как максимум - требует исправления (возможности указать строкой имя элемента для jquery-объекта, например).

**Comment by tadatuta** [comment]
1. `elemParams()` понимает строку в качестве аргумента (см. https://ru.bem.info/libs/bem-core/v2.7.0/desktop/i-bem/jsdoc/#jsdoc-elemParams-1).
2. Ситуация с «кто первый, тот и папа» происходит из-за того, что при передаче элемента как jQuery-цепочки, его имя парсится [вот так](https://github.com/bem/bem-core/blob/v2/common.blocks/i-bem/__dom/i-bem__dom.js#L1449-L1450). Если есть идеи, как его получить иначе — можно сразу завести issue в `bem-core`.

**Comment by apsavin** [comment]
Достаточно разрешить вызов `this.elemParams(elemName, $jqueryObject)`

**Comment by h4** [comment]
@tadatuta мне хочется для начала понять, насколько _bem-way_ размещать несколько элементов на одной дом-ноде и дёргать для них параметры из `data-bem`. Ведь ничто не мешает сделать что-то типа такого:

```html
<span 
    class="post__action post__create" 
    data-bem='{ "post__action": {}, "post__create": {} }'>
    Создать
</span>
```

**Comment by tadatuta** [comment]
@h4 ничего противоречащего bem-way в нескольких элементах не вижу.
Ну и напомню, что в будущей bem-core v3 элементы станут полноправными JS-сущностями и все подобные моменты будут работать изначально правильно.

**Comment by tadatuta** [comment]
@apsavin 

> Достаточно разрешить вызов `this.elemParams(elemName, $jqueryObject)`

Когда есть `elemName` в виде строки, то смысл в `$jqueryObject` теряется. Проблема может быть в том, как из `$(e.target)` вынуть нужный `elemName`.

**Comment by h4** [comment]
@tadatuta я подумаю на досуге, что можно предпринять.

А какой роадмап у bem-core v3?

**Comment by tadatuta** [comment]
@h4 вот [майлстоун](https://github.com/bem/bem-core/milestones/3.0), но скорее всего он будет еще расширяться. А в терминах дат, к сожалению, пока ничего ответить не смогу :(
